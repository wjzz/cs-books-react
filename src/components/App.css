thead tr td {
    font-weight: bold;
    text-align: center;
}

table {
    border: 1px dotted black;
}

tbody td {
    border-top: 1px dotted black;
    border-bottom: 1px dotted black;
}

.cover {
    text-align: right; 
}