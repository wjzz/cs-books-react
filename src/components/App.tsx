import React, {useState} from 'react';
import './App.css';

import {BrowserRouter as Router} from "react-router-dom";

import { getUserBookData, saveUserBookData } from "./Books";
import BookList from './BookList';


const App: React.FC = () => {
  const initial_user_book_data = getUserBookData();
  const [books, setBooks] = useState(initial_user_book_data);

  const handleToggleRead = (index: number) => {
    const newBooks = [...books];
    const bookToChange = books[index];
    newBooks[index] = {...bookToChange, read: !bookToChange.read};
    
    saveUserBookData(newBooks);
    setBooks(newBooks);
  }

  const handleToggleAll = (value: boolean) => {
    const newBooks = books.map(book => ({...book, read: value}));

    saveUserBookData(newBooks);
    setBooks(newBooks);
  }

  const handleSortByTitle = () => {
    const newBooks = [...books].sort((a, b) => a.book.title > b.book.title ? 1 : -1);
    setBooks(newBooks);
  }

  const handleSortByKey = () => {
    const newBooks = [...books].sort((a, b) => a.index - b.index);
    setBooks(newBooks);
  }

  return (
    <Router>
      <BookList 
        books={books}
        toggleRead={handleToggleRead}
        toggleAll={handleToggleAll}
        sortByTitle={handleSortByTitle}
        sortByKey={handleSortByKey}
      />
    </Router>
  );
}

export default App;
