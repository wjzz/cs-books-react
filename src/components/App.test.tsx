import React from "react";
import ReactDOM from "react-dom";

import renderer from "react-test-renderer";
import { mount, render } from "enzyme";

import App from "./App";

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test("App is in sync with the snapshot", () => {
  const component = renderer.create(<App />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("App correctly renders the basic structure", () => {
  const app = render(<App />);

  expect(app.text()).toContain("There are 4 books in total");
  expect(app.find("button").length).toEqual(8);
});

const countChars = (haystack: string, needle: string): number => {
  return Array.from(haystack).filter(char => char === needle).length;
};

test("App correctly reacts to 'mark all' button clicks", () => {
  const app = mount(<App />);

  const unread_marker = "❌";
  const read_marker = "✅";

  app.find("#read-none").simulate("click");

  expect(app.text()).toContain(
    "You have read 0 books, which is 0% of all available books."
  );
  expect(countChars(app.text(), read_marker)).toEqual(0);
  expect(countChars(app.text(), unread_marker)).toEqual(4);

  app.find("#read-all").simulate("click");

  expect(app.text()).toContain(
    "You have read 4 books, which is 100% of all available books."
  );
  expect(countChars(app.text(), read_marker)).toEqual(4);
  expect(countChars(app.text(), unread_marker)).toEqual(0);
});
