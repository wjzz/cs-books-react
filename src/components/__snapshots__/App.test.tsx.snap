// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App is in sync with the snapshot 1`] = `
Array [
  <table>
    <thead>
      <tr>
        <td>
          Cover
        </td>
        <td>
          Info
        </td>
        <td>
          Actions
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="cover"
        >
          <img
            alt="book cover"
            height="80"
            src="/covers/up.jpg"
          />
        </td>
        <td>
          ❌
          <a
            href="/focus/0"
            onClick={[Function]}
          >
            You don't know JS: Up & Going
             
          </a>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
             Mark as 
            read
             
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="cover"
        >
          <img
            alt="book cover"
            height="80"
            src="/covers/sicp.jpg"
          />
        </td>
        <td>
          ❌
          <a
            href="/focus/1"
            onClick={[Function]}
          >
            Structure and Interpretation of Computer Programs
             
          </a>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
             Mark as 
            read
             
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="cover"
        >
          <img
            alt="book cover"
            height="80"
            src="/covers/tapl.jpg"
          />
        </td>
        <td>
          ❌
          <a
            href="/focus/2"
            onClick={[Function]}
          >
            Types and Programming Languages
             
          </a>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
             Mark as 
            read
             
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="cover"
        >
          <img
            alt="book cover"
            height="80"
            src="/covers/eloquent.jpg"
          />
        </td>
        <td>
          ❌
          <a
            href="/focus/3"
            onClick={[Function]}
          >
            Eloquent JavaScript
             
          </a>
        </td>
        <td>
          <button
            onClick={[Function]}
          >
             Mark as 
            read
             
          </button>
        </td>
      </tr>
    </tbody>
  </table>,
  <div>
    There are 4 books in total.
     
    <br />
    You have read 0 books, which is 0% of all available books.
    <br />
    <button
      id="read-none"
      onClick={[Function]}
    >
      Mark all as unread
    </button>
    <span>
       
    </span>
    <button
      id="read-all"
      onClick={[Function]}
    >
      Mark all as read
    </button>
    <br />
    <button
      onClick={[Function]}
    >
      Sort by title
    </button>
    <span>
       
    </span>
    <button
      onClick={[Function]}
    >
      Sort by key
    </button>
  </div>,
]
`;
